### DISCUSSION TITLE
Is it possible to put 'util/library' shell scripts in `.chezmoiscripts` dir?

### DISCUSSION THREAD

**Discussion by kquinsland** [twpayne/chezmoi#3506] [Q&A] [answered]
I have a series of scripts that should run _before_ chezmoi begins to render templates (think "install the `op` cli tool and `1password` desktop app") and some scripts that should be run after certain files are changed (think "rebuild `zgenom` files after `~/.zshrc` updated")

Here's what things look like:

```shell
# From inside source root:
❯ tree .chezmoiscripts
.chezmoiscripts
├── run_after_dot_zshrc.sh
├── run_before_00-install-pre-requisites.sh
├── run_once_before_install-fonts.sh
├── run_once_before_install-zsh-starter-kit.sh
├── run_onchange_install-packages.sh.tmpl
└── scripts
    ├── install-1password-cli.sh
    └── utils.sh
```

Where `utils.sh` contains your typical "kitchen-sink" of helper functions to install packages, log stuff, backup/symlink stuff ...etc.
The `install-1password-cli.sh` script is meant to be run independentally / directly or `source`'d from another script (` run_before_00-install-pre-requisites.sh` does just that: `source scripts/install-1password-cli.sh`)

When I run `chezmoi apply`, I get this output:

```shell
❯ cz apply
chezmoi: $mySourceDirPath/home/.chezmoiscripts/scripts/utils.sh: not a script
(this repeats for `install-1password-cli.sh` as well...)
```

Which i've tracked down to [here](https://github.com/twpayne/chezmoi/blob/f0a7529aab244e3fc8d3439dab867c380f6b7cdb/internal/chezmoi/sourcestate.go#L2668) and the error is happening [because the script does not have a `run_` prefix](https://github.com/twpayne/chezmoi/blob/master/internal/chezmoi/attr.go#L180C31-L180C40).

**My question**

Is there an incantation I can add to `.chezmoiignore` or similar so that I can keep helper/util/library scripts co-located with with the chezmoi scripts or am I going to be forced to put only `run_*` scripts in `.chezmoiscripts` and locate util/library code elsewhere?

The reason i'd like to avoid re-locating things: relative paths won't really work given the `/tmp` dir where `chezmoi` renders things out:

```shell
/tmp/3963292091.00-install-pre-requisites.sh: line 12: ../../scripts/utils.sh: No such file or directory
```

**Comment by twpayne** [comment]
> I have a series of scripts that should run _before_ chezmoi begins to render templates (think "install the `op` cli tool and `1password` desktop app") and some scripts that should be run after certain files are changed (think "rebuild `zgenom` files after `~/.zshrc` updated")

chezmoi has `run_once_before_` (for installing your password manager) and `run_onchange_after_` scripts for exactly this sort of thing. You can use [this trick](https://www.chezmoi.io/user-guide/use-scripts-to-perform-actions/#run-a-script-when-the-contents-of-another-file-changes) for rebuilding when `~/.zshrc` is updated. Can you use these?

**Reply by kquinsland** [reply]
Thanks for a super quick reply!

That's what the scripts are meant to do but they're pretty modular/composable. E.g.: here's one of the functions that lives in `utils.sh`:

```shell
install_binary() {
    ##
    # Simple abstraction to install a binary using a supported package manager
    # Example usage:
    # install_binary "fzf"
    # install_binary "fzf" "dnf"

    local package_name="${1}"
    local package_manager="${2}"

    # Function to determine the package manager if not specified
    determine_package_manager() {
        if command -v apt &>/dev/null; then
            echo "apt"
        elif command -v dnf &>/dev/null; then
            echo "dnf"
        elif command -v pacman &>/dev/null; then
            echo "pacman"
        elif command -v brew &>/dev/null; then
            echo "brew"
        else
            log_message "ERROR" "No supported package manager found."
            exit 1
        fi
    }

    # If the package manager is not specified, determine it
    if [[ -z "${package_manager}" ]]; then
        package_manager=$(determine_package_manager)
    fi

    # Install the package using the determined or specified package manager
    log_message "INFO" "Installing ${package_name} using ${package_manager}..."

    case "${package_manager}" in
    apt)
        sudo apt install -y "${package_name}" || {
            log_message "ERROR" "Installation failed."
            exit 1
        }
        ;;
    dnf)
        sudo dnf install -y "${package_name}" || {
            log_message "ERROR" "Installation failed."
            exit 1
        }
        ;;
    pacman)
        sudo pacman -S "${package_name}" || {
            log_message "ERROR" "Installation failed."
            exit 1
        }
        ;;
    brew)
        brew install "${package_name}" || {
            log_message "ERROR" "Installation failed."
            exit 1
        }
        ;;
    *)
        log_message "ERROR" "Unsupported package manager: ${package_manager}"
        exit 1
        ;;
    esac
    log_message "INFO" "Installation of ${package_name} completed successfully."
}
```

The `install_binary` is something that I use from a few of the `run_*` scripts which is why it's been pulled out into it's own lib.

The issue that I'm running into is that chezmoi gets upset that there's a shell script in `.chezmoiscripts` that does not conform to the `run_*` pattern. My question is asking if there's a way to tell chezmoi to ignore these "libraries" since they're only used by the shell scripts that chezmoi is meant to run.

**Comment by Lockszmith-GH** [comment]
If these scripts are intended to never be copied to your home directory, you can use [`.chezmoiroot`](https://www.chezmoi.io/reference/special-files-and-directories/chezmoiroot/) to move your home-dir content, allowing your repo to contain files that will not be copied over.

```text
<git-repo-root>
├── .chemoihome # pointing at my_home
├── my_home
|   └── .chezmoiscripts
|       ├── run_after_dot_zshrc.sh
|       ├── run_before_00-install-pre-requisites.sh
|       ├── run_once_before_install-fonts.sh
|       ├── run_once_before_install-zsh-starter-kit.sh
|       └── run_onchange_install-packages.sh.tmpl
└── scripts
    ├── install-1password-cli.sh
    └── utils.sh
```

**Reply by kquinsland** [reply, marked as answer]
Thanks for the reply!

I am already using the chezmoiroot trick :) and your proposed layout is exactly what I tried before opening this thread. Unfortunately, it doesn't work since each `run_*.sh` has something to the effect of `source ../../scripts/utils.sh` at the top but this relative file include path breaks when chezmoi copies the `run_*` scripts:

```shell
❯ cz apply
/tmp/3963292091.00-install-pre-requisites.sh: line 12: ../../scripts/utils.sh: No such file or directory
chezmoi: exit status 1
```

Versus when the `scripts/utils.sh` lives inside of .chezmoiscripts`:

```shell
❯ cz apply
chezmoi: $mySourceDirPath/home/.chezmoiscripts/scripts/utils.sh: not a script
```


-----


I was able to work around the issue by changing the `source` directive so it becomes a fully qualified file path when `chezmoi` runs the scripts:

```shell
source ${CHEZMOI_SOURCE_DIR}/../scripts/utils.sh
```

**Reply by halostatue** [reply]
Using `${CHEZMOI_SOURCE_DIR}` (this is `.chezmoiroot`) or `${CHEZMOI_WORKING_TREE}` (that skips the need for the `../`) is the correct solution to have a shared library script that can be used.

If you do not use `.chezmoiroot`, you can get the same effect by having a `.lib` (or whatever, but it must be prefixed by `.`) directory in your working tree and then you `source "${CHEZMOI_SOURCE_DIR}/.lib/util.sh"`.

**Reply by kquinsland** [reply]
Thanks for clarifying and the tip about working_tree!

For completeness / the next guy that lands here after a google search, here's the bit that does the trick:

```shell
# -u barfs on unset vars
set -eu

if [[ -z "${CHEZMOI_WORKING_TREE}" ]]; then
    echo "Error, \$CHEZMOI_WORKING_TREE must be set. Please only run this script via chezmoi"
    exit 1
fi
source "${CHEZMOI_WORKING_TREE}/scripts/utils.sh"

# First up, install 1password as it's the most important dependency
source ${CHEZMOI_WORKING_TREE}/scripts/install-1password-cli.sh

...
```

**Reply by Lockszmith-GH** [reply]
Just some bash syntactic sugar - a trick I don't see utilized enough, which funcitons pretty much the same:

~As you've `set -e`,~ remove the `if` statement and replace the `source` line with
```bash
source "${CHEZMOI_WORKING_TREE:?env variable missing. Please only run this script via chezmoi}/scripts/utils.sh"
```

When CHEZMOI_WORKING_TREE is missing/empty, the exit-code will be `1` and the following error will come up:
> bash: CHEZMOI_WORKING_TREE: env variable missing. Please only run this script via chezmoi

**Reply by halostatue** [reply]
`set -e` is not required for `${VAR:?message}` to cause the script to halt:

> `${parameter:?word}`  Display Error if Null or Unset.  If parameter is null or unset, the expansion of word (or a message to that effect if word is not present) is written to the standard error and the shell, if it is not interactive, exits.  Otherwise, the value of parameter is substituted.

**Reply by shermanjoshua** [reply]
> Thanks for clarifying and the tip about working_tree!
> 
> For completeness / the next guy that lands here after a google search, here's the bit that does the trick:
> 
> ```shell
> # -u barfs on unset vars
> set -eu
> 
> if [[ -z "${CHEZMOI_WORKING_TREE}" ]]; then
>     echo "Error, \$CHEZMOI_WORKING_TREE must be set. Please only run this script via chezmoi"
>     exit 1
> fi
> source "${CHEZMOI_WORKING_TREE}/scripts/utils.sh"
> 
> # First up, install 1password as it's the most important dependency
> source ${CHEZMOI_WORKING_TREE}/scripts/install-1password-cli.sh
> 
> ...
> ```

🙋 Next person here! This wasn't a direct fix for me, and I haven't fully solved yet, but I'm on the right track now and making progress. So I'm making a pit-stop now and dropping it to say thanks to everyone here for the help. (I always mean to do this, but I never stop to leave the comment. Trying to do better 😄) 

I'm actually hitting this error not from inside another script, but referencing directly from my `.chezmoi.yml.tmpl` and by extension the yaml file that it generates. I was using this to build the path that was getting me "not a script" errors (I also have a `.chezmoiroot` file with src as the directory to redirect to..... the re-directory???):

```
hooks:
  read-source-state:
    pre:
      command: {{ .chezmoi.workingTree }}/src/.chezmoiscripts/install-password-manager.sh
```

So swapping that out with a fq path does handle the error... kind of. I applied the change to the template and init-ed it. The change does appear in the actual yml file now, but I still get that error in the console. The change is that now - I know I'm hitting the script too, as I've done the ole `echo "iidjafkhaja"` trick and I DO see that in the console in addition to the error. 

So to me, this screams that I'm actually hitting another file with that path in it that needs updating. Or maybe a cache entry somewhere or something? I'm not sure but I'll figure it out sometime. Point is - thanks for getting me on the right track! And if anyone has thoughts, of course I'll give it a shot and report back.

![image](https://github.com/user-attachments/assets/a4389591-f6c4-44e8-8899-4ac0e1d812c5)

**Reply by Lockszmith-GH** [reply]
did you chmod +x the shell script?

instead of just providing the script path, you can try calling `bash <path>`
